pkg/chain/ethereum/ethutil/rate_limiter_test.go

```diff
@@ -0,0 +1,350 @@
+				_, err := rateLimitingBackend.EstimateGas(
+					context.Background(),
+					ethereum.CallMsg{},
+				)
+				return err
+			},
+		},
+		"test SendTransaction rate limiting": {
+			function: func() error {
+				err := rateLimitingBackend.SendTransaction(
+					context.Background(),
+					nil,
+				)
+				return err
+			},
+		},
+		"test FilterLogs rate limiting": {
+			function: func() error {
+				_, err := rateLimitingBackend.FilterLogs(
+					context.Background(),
+					ethereum.FilterQuery{},
+				)
+				return err
+			},
+		},
+		"test SubscribeFilterLogs rate limiting": {
+			function: func() error {
+				_, err := rateLimitingBackend.SubscribeFilterLogs(
+					context.Background(),
+					ethereum.FilterQuery{},
+					nil,
+				)
+				return err
+			},
+		},
+	}
+
+	for testName, test := range tests {
+		t.Run(testName, func(t *testing.T) {
+			wg := sync.WaitGroup{}
+			wg.Add(requests)
+
+			startSignal := make(chan struct{})
+
+			for i := 0; i < requests; i++ {
+				go func() {
+					<-startSignal
+
+					err := test.function()
+					if err != nil {
+						t.Errorf("unexpected error: [%v]", err)
+					}
+
+					wg.Done()
+				}()
+			}
+
+			startTime := time.Now()
+			close(startSignal)
+
+			wg.Wait()
+
+			duration := time.Now().Sub(startTime)
+			averageRequestsPerSecond := float64(requests) / duration.Seconds()
+
+			if averageRequestsPerSecond > float64(requestsPerSecondLimit) {
+				t.Errorf(
+					"average requests per second exceeded the limit\n"+
+						"limit:  [%v]\n"+
+						"actual: [%v]",
+					requestsPerSecondLimit,
+					averageRequestsPerSecond,
+				)
+			}
+
+			maxConcurrency := 0
+			temporaryConcurrency := 0
+			for _, event := range backend.events {
+				if event == "start" {
+					temporaryConcurrency++
+				}
+
+				if event == "end" {
+					temporaryConcurrency--
+				}
+
+				if temporaryConcurrency > maxConcurrency {
+					maxConcurrency = temporaryConcurrency
+				}
+			}
+
+			if maxConcurrency > concurrencyLimit {
+				t.Errorf(
+					"max concurrency exceeded the limit\n"+
+						"limit:  [%v]\n"+
+						"actual: [%v]",
+					concurrencyLimit,
+					maxConcurrency,
+				)
+			}
+		})
+	}
+}
+
+func TestRateLimiter_AcquirePermitTimout(t *testing.T) {
+	requestsPerSecondLimit := 1
+	concurrencyLimit := 1
+	acquirePermitTimeout := 10 * time.Millisecond
+	requests := 3
+	requestDuration := 250 * time.Millisecond
+
+	backend := &mockBackend{
+		requestDuration,
+		make([]string, 0),
+		sync.Mutex{},
+	}
+
+	rateLimitingBackend := WrapRateLimiting(
+		backend,
+		&RateLimiterConfig{
+			RequestsPerSecondLimit: requestsPerSecondLimit,
+			ConcurrencyLimit:       concurrencyLimit,
+			AcquirePermitTimeout:   acquirePermitTimeout,
+		},
+	)
+
+	wg := sync.WaitGroup{}
+	wg.Add(requests)
+
+	startSignal := make(chan struct{})
+	errorChannel := make(chan error, requests)
+	errors := make([]error, 0)
+
+	go func() {
+		for e := range errorChannel {
+			errors = append(errors, e)
+		}
+	}()
```

Comment:
Instead of reading from the channel we can give this channel a buffer of `requests` - I think it's a little cleaner/easier to understand.

Reply:
It is buffered.

Reply:
Ah, right. I missed that `0` is passed to `errors` initialization call and not to `errorChannel` initialization call. Neverthless, having the buffered channel we can get rid of the entire
```
go func() {
    for e := range errorChannel {
        errors = append(errors, e)
    }
}()
```
goroutine and do:
```
errors := make(chan error, requests)
// (...)

close(errors)
if len(errors) == 0 {
  // (...)
}

for e := range errors {
  // (...)
}
```

By using the buffer, we are not leaving any goroutine from test dangling.

Reply:
Ah got your point. Indeed we can simplify this. I will change it.

Reply:
Done.